File: bindings/go/tests/test_async_extractor_test.go
package tests

import (
	"context"
	"testing"

	tiktokdirect "github.com/codewithwan/tiktok-direct-go"
)

func TestInvalidUrlRaisesErrorAsync(t *testing.T) {
	extractor := tiktokdirect.New()
	
	errChan := make(chan error, 1)
	go func() {
		_, err := extractor.Extract(context.Background(), "invalid-url")
		errChan <- err
	}()
	
	err := <-errChan
	if err == nil {
		t.Error("Expected error for invalid url asynchronously, got nil")
	}
}

func TestExtractRealUrlAsync(t *testing.T) {
	extractor := tiktokdirect.New()
	
	type result struct {
		meta map[string]any
		err  error
	}
	resChan := make(chan result, 1)
	
	go func() {
		meta, err := extractor.Extract(context.Background(), URL)
		resChan <- result{meta, err}
	}()
	
	res := <-resChan
	if res.err != nil {
		t.Fatalf("Failed to extract real URL async: %v", res.err)
	}

	if res.meta["quality"] != "complete" && res.meta["quality"] != "partial" {
		t.Errorf("Expected complete or partial, got %v", res.meta["quality"])
	}
	if fmtID := res.meta["video_id"]; fmtID == nil || fmtID == "" {
		t.Error("Expected video_id, got empty")
	}
	if fmtTitle := res.meta["title"]; fmtTitle == nil || fmtTitle == "" {
		t.Error("Expected title, got empty")
	}
}

File: bindings/node/tests/test_async_extractor.test.js
import assert from "node:assert/strict";
import { test } from "node:test";
import { TikTokExtractor, extract } from "../src/index.js";

const URL = "https://vt.tiktok.com/ZSxvYRvoR/";

test("test_invalid_url_raises_error_async", async () => {
  const extractor = new TikTokExtractor();
  await assert.rejects(
    async () => await extractor.extract("invalid-url")
  );

  await assert.rejects(
    async () => await extract("invalid-url")
  );
});

test("test_extract_real_url_async", async () => {
  const extractor = new TikTokExtractor();
  const meta = await extractor.extract(URL);

  assert.equal(meta.quality, "complete");
  assert.ok(meta.video_id);
  assert.ok(meta.title);
});
